src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs

```diff
@@ -56,22 +78,74 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
+                        _module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__GetPinnableReference) is MethodSymbol getPinnableReference)
                     {
-                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
+                        // Use RuntimeHelpers.CreateSpan and cpblk.
+                        EmitStackAllocBlockMultiByteInitializer(data, createSpanHelper, getPinnableReference, elementType, inits.Syntax, _diagnostics.DiagnosticBag);
+                    }
+                    else
+                    {
+                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                        mixedInitialized = true;
                     }
                 }
-                else
+
+                if (initializationStyle == ArrayInitializerStyle.Mixed && !mixedInitialized)
                 {
-                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
                 }
             }
+
+            void emitLocalloc()
+            {
+                EmitExpression(count, used: true);
+
+                _sawStackalloc = true;
+                _builder.EmitOpCode(ILOpCode.Localloc);
+            }
+
+            static Cci.IMethodReference? getCreateSpanHelper(Emit.PEModuleBuilder module, TypeSymbol elementType)
+            {
+                var member = module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+                return ((MethodSymbol?)member)?.Construct(elementType).GetCciAdapter();
+            }
+        }
+
+        private void EmitStackAllocBlockMultiByteInitializer(ImmutableArray<byte> data, Cci.IMethodReference createSpanHelper, MethodSymbol getPinnableReferenceDefinition, TypeSymbol elementType, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        {
+            var readOnlySpan = getPinnableReferenceDefinition.ContainingType.Construct(elementType);
+            Debug.Assert(TypeSymbol.Equals(readOnlySpan.OriginalDefinition, _module.Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.ConsiderEverything));
+            var getPinnableReference = getPinnableReferenceDefinition.AsMember(readOnlySpan);
+
+            _builder.EmitOpCode(ILOpCode.Dup);
+            _builder.EmitCreateSpan(data, createSpanHelper, syntaxNode, diagnostics);
+
+            var temp = AllocateTemp(readOnlySpan, syntaxNode);
+            _builder.EmitLocalStore(temp);
+            _builder.EmitLocalAddress(temp);
+
+            _builder.EmitOpCode(ILOpCode.Call, 0);
+            EmitSymbolToken(getPinnableReference, syntaxNode, optArgList: null);
+            _builder.EmitIntConstant(data.Length);
+            _builder.EmitOpCode(ILOpCode.Cpblk, -3);
```

Comment:
>_builder.EmitOpCode(ILOpCode.Cpblk, -3); [](http://example.com/codeflow?start=12&length=40)

ECMA-335 says:
> cpblk assumes that both destaddr and srcaddr are aligned to the natural size of the machine

Are we ensuring that this requirement is met?

Reply:
> and the field has a structure type, so I think it should be aligned properly.

A blob of bytes is not an instance of a structure type. We are not copying data from a field.

Reply:
I thought we are emitting a special struct (e.g., `__StaticArrayInitTypeSize=12`) and a field in `<PrivateImplementationDetails>` which is of the type of that struct, initialized with a blob of data, and we are using that field - so that field should be naturally aligned as it's of the struct type. Plus `CreateSpan` can return pointer to a different blob of data with fixed-up endianness, but it should be also naturally aligned as it allocates (and caches) an array in that case.

Reply:
@stephentoub can you advise how we ensure we meet `cpblk`'s alignment requirements here?

Btw, looks like this code creates an RVA field with alignment 1 and uses it with `cpblk`, which, if correct, would suggest `cpblk` doesn't need any special alignment of the field:

https://github.com/dotnet/roslyn/blob/67ef2fb7409e0ab54ac627d6ac525069172ed494/src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs#L50-L58

Reply:
You can prefix the cpblk instruction with unaligned in order to avoid the alignment requirement, which IIRC wants the pointers to be aligned to the machine's natural size.  It's possible the cited existing code should be using unaligned as well, or maybe the requirement is less about the machine's natural word size and more about the size of the type in use (or maybe this is a place where ECMA and coreclr differ). @jkotas?

Reply:
If you would like to be compliant with ECMA, emit the `unaligned` prefix. It is not required to make things work on current runtimes. `cpblk` implementations in current runtimes do not make any assumptions about the alignment of the inputs.